File: Ch. 12/Exhibit 12.1.1A/code/exhibit_12_1_1A.py
# =============================================================================
# Exhibit 12.1.1A: ATEs With and Without Available Case Analysis
# =============================================================================
# Compares treatment effects with and without available case analysis.
# Columns 1-2: Without controls
# Columns 3-4: With controls (female, race_w, hl_eng_span, birthweight)
#
# Reference: Chapter 12, Addressing Attrition

from pathlib import Path
import pandas as pd
import statsmodels.api as sm


# --- Setup -------------------------------------------------------------------
# Automatically resolve paths relative to this script's location.
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR.parent / "data"
OUTPUT_DIR = SCRIPT_DIR.parent / "output"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


# --- Load and filter data ----------------------------------------------------
# Load synthetic data (for testing/demonstration)
data_12_1_1A = pd.read_stata(DATA_DIR / "unique_data_clean_main_synthetic.dta")
# To use actual CHECC data, comment out the line above and uncomment the line below:
# data_12_1_1A = pd.read_stata(DATA_DIR / "unique_data_clean_main.dta")


data_12_1_1A['year'] = pd.to_numeric(data_12_1_1A['year'], errors='coerce')

# Apply filters: year >= 2012, treatment in {control, prek}, exclude kinderprep
# and late randomized, require baseline cognitive score
cleaned_data_12_1_1A = data_12_1_1A[
    (data_12_1_1A['year'] >= 2012) &
    ((data_12_1_1A['treatment'] == 'control') |
     (data_12_1_1A['treatment'] == 'prek')) &
    (data_12_1_1A['kinderprep'] == 0) &
    (data_12_1_1A['late_randomized'] == 0) &
    (data_12_1_1A['has_cog_pre'] != 0)
].copy()

# Create block variable (prioritize 2012 block, fallback to 2013)
cleaned_data_12_1_1A['block'] = cleaned_data_12_1_1A.apply(
    lambda row: row['block_2012'] if row['block_2012'] != "" else row['block_2013'],
    axis=1
)

# Set summer loss cognitive score to NA when not observed
cleaned_data_12_1_1A['std_cog_sl'] = cleaned_data_12_1_1A['std_cog_sl'].where(
    cleaned_data_12_1_1A['has_cog_sl'] != 0, pd.NA
)

# Treatment indicator: 1 if pre-K, 0 if control
cleaned_data_12_1_1A['d_i'] = (cleaned_data_12_1_1A['treatment'] == "prek").astype(int)

# Response indicator: 1 if outcome observed, 0 if attrited
cleaned_data_12_1_1A['r_i'] = cleaned_data_12_1_1A['std_cog_sl'].notna().astype(int)


# --- Filter for available case analysis -------------------------------------
# Available case analysis: restrict to observations with observed outcomes
filtered_data_12_1_1A = cleaned_data_12_1_1A[cleaned_data_12_1_1A['r_i'] == 1]


# --- Estimate models ---------------------------------------------------------
# Column 1: Default model without controls (full data, includes missing outcomes)
model_default = sm.OLS.from_formula(
    'std_cog_sl ~ d_i',
    data=cleaned_data_12_1_1A
).fit()

# Column 2: Available case analysis without controls (only observed outcomes)
model_aca = sm.OLS.from_formula(
    'std_cog_sl ~ d_i',
    data=filtered_data_12_1_1A
).fit()

# Column 3: Default model with controls (full data)
model_default_controls = sm.OLS.from_formula(
    'std_cog_sl ~ d_i + female + race_w + hl_eng_span + birthweight',
    data=cleaned_data_12_1_1A
).fit()

# Column 4: Available case analysis with controls (only observed outcomes)
model_aca_controls = sm.OLS.from_formula(
    'std_cog_sl ~ d_i + female + race_w + hl_eng_span + birthweight',
    data=filtered_data_12_1_1A
).fit()


# --- Print results -----------------------------------------------------------
print("\n" + "=" * 80)
print("EXHIBIT 12.1.1A: ATEs With and Without Available Case Analysis")
print("=" * 80)
print(f"{'':20} {'(1)':>15} {'(2)':>15} {'(3)':>15} {'(4)':>15}")
print(f"{'':20} {'Default':>15} {'Available':>15} {'Default':>15} {'Available':>15}")
print(f"{'':20} {'Model':>15} {'Case':>15} {'Model':>15} {'Case':>15}")
print("-" * 80)

# Pre-K coefficient row
print(f"{'Pre-K':<20} {model_default.params['d_i']:>15.3f} "
      f"{model_aca.params['d_i']:>15.3f} "
      f"{model_default_controls.params['d_i']:>15.3f} "
      f"{model_aca_controls.params['d_i']:>15.3f}")
print(f"{'':20} ({model_default.bse['d_i']:>13.3f}) "
      f"({model_aca.bse['d_i']:>13.3f}) "
      f"({model_default_controls.bse['d_i']:>13.3f}) "
      f"({model_aca_controls.bse['d_i']:>13.3f})")

# Constant row
print(f"{'Constant':<20} {model_default.params['Intercept']:>15.3f} "
      f"{model_aca.params['Intercept']:>15.3f} "
      f"{model_default_controls.params['Intercept']:>15.3f} "
      f"{model_aca_controls.params['Intercept']:>15.3f}")
print(f"{'':20} ({model_default.bse['Intercept']:>13.3f}) "
      f"({model_aca.bse['Intercept']:>13.3f}) "
      f"({model_default_controls.bse['Intercept']:>13.3f}) "
      f"({model_aca_controls.bse['Intercept']:>13.3f})")

# Controls indicator row
print(f"{'Controls':<20} {'No':>15} {'No':>15} {'Yes':>15} {'Yes':>15}")

# R-squared row
print(f"{'R-squared':<20} {model_default.rsquared:>15.3f} "
      f"{model_aca.rsquared:>15.3f} "
      f"{model_default_controls.rsquared:>15.3f} "
      f"{model_aca_controls.rsquared:>15.3f}")

# Observations row
print(f"{'Observations':<20} {int(model_default.nobs):>15} "
      f"{int(model_aca.nobs):>15} "
      f"{int(model_default_controls.nobs):>15} "
      f"{int(model_aca_controls.nobs):>15}")
print("=" * 80)


# --- Save results to LaTeX ---------------------------------------------------
# Create summary table with regression results
results_table = pd.DataFrame({
    'Variable': ['Pre-K', '', 'Constant', '', 'Controls', 'R-squared', 'Observations'],
    '(1) Default Model': [
        f"{model_default.params['d_i']:.3f}",
        f"({model_default.bse['d_i']:.3f})",
        f"{model_default.params['Intercept']:.3f}",
        f"({model_default.bse['Intercept']:.3f})",
        'No',
        f"{model_default.rsquared:.3f}",
        f"{int(model_default.nobs)}"
    ],
    '(2) Available Case': [
        f"{model_aca.params['d_i']:.3f}",
        f"({model_aca.bse['d_i']:.3f})",
        f"{model_aca.params['Intercept']:.3f}",
        f"({model_aca.bse['Intercept']:.3f})",
        'No',
        f"{model_aca.rsquared:.3f}",
        f"{int(model_aca.nobs)}"
    ],
    '(3) Default Model': [
        f"{model_default_controls.params['d_i']:.3f}",
        f"({model_default_controls.bse['d_i']:.3f})",
        f"{model_default_controls.params['Intercept']:.3f}",
        f"({model_default_controls.bse['Intercept']:.3f})",
        'Yes',
        f"{model_default_controls.rsquared:.3f}",
        f"{int(model_default_controls.nobs)}"
    ],
    '(4) Available Case': [
        f"{model_aca_controls.params['d_i']:.3f}",
        f"({model_aca_controls.bse['d_i']:.3f})",
        f"{model_aca_controls.params['Intercept']:.3f}",
        f"({model_aca_controls.bse['Intercept']:.3f})",
        'Yes',
        f"{model_aca_controls.rsquared:.3f}",
        f"{int(model_aca_controls.nobs)}"
    ]
})

# Save to LaTeX
tex_file = OUTPUT_DIR / "Exhibit_12_1_1A_python.tex"
results_table.to_latex(tex_file, index=False, escape=False,
                       caption="ATEs With and Without Available Case Analysis",
                       label="tab:exhibit_12_1_1A")
print(f"\n✓ Saved to: {tex_file}\n")

# =============================================================================
# END OF EXHIBIT 12.1.1A
# =============================================================================

File: Ch. 12/Exhibit 12.1.1A/code/exhibit_12_1_1A.R
# =============================================================================
# Exhibit 12.1.1A: ATEs With and Without Available Case Analysis
# =============================================================================
# Compares treatment effects with and without available case analysis.
# Columns 1-2: Without controls
# Columns 3-4: With controls (female, race_w, hl_eng_span, birthweight)
#
# Reference: Chapter 12, Addressing Attrition

# --- Setup -------------------------------------------------------------------
rm(list = ls())

# Load required libraries
library(tidyverse)
library(haven)
library(xtable)

# Automatically set working directory to the folder containing this script.
# Supports RStudio (Source/Knit), source(), and Rscript from the command line.
if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
} else if (length(sys.frames()) > 0 && !is.null(sys.frame(1)$ofile)) {
  setwd(dirname(sys.frame(1)$ofile))
} else if (!is.null(commandArgs(trailingOnly = FALSE))) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    setwd(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
}

# Define paths relative to script location
data_dir <- file.path(dirname(getwd()), "data")
output_dir <- file.path(dirname(getwd()), "output")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


# --- Load and filter data ----------------------------------------------------
# Load synthetic data (for testing/demonstration)
data <- read_dta(file.path(data_dir, "unique_data_clean_main_synthetic.dta"))
# To use actual CHECC data, comment out the line above and uncomment the line below:
# data <- read_dta(file.path(data_dir, "unique_data_clean_main.dta"))

# Apply filters and create necessary variables
cleaned_data <- data %>%
  filter(year >= 2012 &
         (treatment == "control" | treatment == "prek") &
         kinderprep == 0 &
         late_randomized == 0 &
         has_cog_pre != 0) %>%
  mutate(
    # Create block variable (prioritize 2012 block, fallback to 2013)
    block = case_when(
      block_2012 != "" ~ block_2012,
      block_2013 != "" ~ block_2013
    ),
    # Set summer loss cognitive score to NA when not observed
    std_cog_sl = ifelse(has_cog_sl == 0, NA, std_cog_sl),
    # Treatment indicator: 1 if pre-K, 0 if control
    d_i = (treatment == "prek"),
    # Response indicator: 1 if outcome observed, 0 if attrited
    r_i = !is.na(std_cog_sl)
  )


# --- Filter for available case analysis -------------------------------------
# Available case analysis: restrict to observations with observed outcomes
filtered_data <- cleaned_data %>% filter(r_i == 1)


# --- Estimate models ---------------------------------------------------------
# Column 1: Default model without controls (full data, includes missing outcomes)
model_default <- lm(
  std_cog_sl ~ d_i,
  data = cleaned_data
)

# Column 2: Available case analysis without controls (only observed outcomes)
aca.model <- lm(
  std_cog_sl ~ d_i,
  data = filtered_data
)

# Column 3: Default model with controls (full data)
model_default_controls <- lm(
  std_cog_sl ~ d_i + female + race_w + hl_eng_span + birthweight,
  data = cleaned_data
)

# Column 4: Available case analysis with controls (only observed outcomes)
aca.model.controls <- lm(
  std_cog_sl ~ d_i + female + race_w + hl_eng_span + birthweight,
  data = filtered_data
)


# --- Print results -----------------------------------------------------------
cat("\n", strrep("=", 80), "\n", sep = "")
cat("EXHIBIT 12.1.1A: ATEs With and Without Available Case Analysis\n")
cat(strrep("=", 80), "\n", sep = "")
cat(sprintf("%-20s %15s %15s %15s %15s\n", "", "(1)", "(2)", "(3)", "(4)"))
cat(sprintf("%-20s %15s %15s %15s %15s\n", "", "Default", "Available", "Default", "Available"))
cat(sprintf("%-20s %15s %15s %15s %15s\n", "", "Model", "Case", "Model", "Case"))
cat(strrep("-", 80), "\n", sep = "")

# Pre-K coefficient row
cat(sprintf("%-20s %15.3f %15.3f %15.3f %15.3f\n",
            "Pre-K",
            coef(model_default)["d_iTRUE"],
            coef(aca.model)["d_iTRUE"],
            coef(model_default_controls)["d_iTRUE"],
            coef(aca.model.controls)["d_iTRUE"]))
cat(sprintf("%-20s (%13.3f) (%13.3f) (%13.3f) (%13.3f)\n",
            "",
            summary(model_default)$coefficients["d_iTRUE", "Std. Error"],
            summary(aca.model)$coefficients["d_iTRUE", "Std. Error"],
            summary(model_default_controls)$coefficients["d_iTRUE", "Std. Error"],
            summary(aca.model.controls)$coefficients["d_iTRUE", "Std. Error"]))

# Constant row
cat(sprintf("%-20s %15.3f %15.3f %15.3f %15.3f\n",
            "Constant",
            coef(model_default)["(Intercept)"],
            coef(aca.model)["(Intercept)"],
            coef(model_default_controls)["(Intercept)"],
            coef(aca.model.controls)["(Intercept)"]))
cat(sprintf("%-20s (%13.3f) (%13.3f) (%13.3f) (%13.3f)\n",
            "",
            summary(model_default)$coefficients["(Intercept)", "Std. Error"],
            summary(aca.model)$coefficients["(Intercept)", "Std. Error"],
            summary(model_default_controls)$coefficients["(Intercept)", "Std. Error"],
            summary(aca.model.controls)$coefficients["(Intercept)", "Std. Error"]))

# Controls indicator row
cat(sprintf("%-20s %15s %15s %15s %15s\n", "Controls", "No", "No", "Yes", "Yes"))

# R-squared row
cat(sprintf("%-20s %15.3f %15.3f %15.3f %15.3f\n",
            "R-squared",
            summary(model_default)$r.squared,
            summary(aca.model)$r.squared,
            summary(model_default_controls)$r.squared,
            summary(aca.model.controls)$r.squared))

# Observations row
cat(sprintf("%-20s %15d %15d %15d %15d\n",
            "Observations",
            nobs(model_default),
            nobs(aca.model),
            nobs(model_default_controls),
            nobs(aca.model.controls)))
cat(strrep("=", 80), "\n", sep = "")


# --- Save results to LaTeX ---------------------------------------------------
# Create summary table with regression results
results_table <- data.frame(
  Variable = c("Pre-K", "", "Constant", "", "Controls", "R-squared", "Observations"),
  `(1) Default Model` = c(
    sprintf("%.3f", coef(model_default)["d_iTRUE"]),
    sprintf("(%.3f)", summary(model_default)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(model_default)["(Intercept)"]),
    sprintf("(%.3f)", summary(model_default)$coefficients["(Intercept)", "Std. Error"]),
    "No",
    sprintf("%.3f", summary(model_default)$r.squared),
    sprintf("%d", nobs(model_default))
  ),
  `(2) Available Case` = c(
    sprintf("%.3f", coef(aca.model)["d_iTRUE"]),
    sprintf("(%.3f)", summary(aca.model)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(aca.model)["(Intercept)"]),
    sprintf("(%.3f)", summary(aca.model)$coefficients["(Intercept)", "Std. Error"]),
    "No",
    sprintf("%.3f", summary(aca.model)$r.squared),
    sprintf("%d", nobs(aca.model))
  ),
  `(3) Default Model` = c(
    sprintf("%.3f", coef(model_default_controls)["d_iTRUE"]),
    sprintf("(%.3f)", summary(model_default_controls)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(model_default_controls)["(Intercept)"]),
    sprintf("(%.3f)", summary(model_default_controls)$coefficients["(Intercept)", "Std. Error"]),
    "Yes",
    sprintf("%.3f", summary(model_default_controls)$r.squared),
    sprintf("%d", nobs(model_default_controls))
  ),
  `(4) Available Case` = c(
    sprintf("%.3f", coef(aca.model.controls)["d_iTRUE"]),
    sprintf("(%.3f)", summary(aca.model.controls)$coefficients["d_iTRUE", "Std. Error"]),
    sprintf("%.3f", coef(aca.model.controls)["(Intercept)"]),
    sprintf("(%.3f)", summary(aca.model.controls)$coefficients["(Intercept)", "Std. Error"]),
    "Yes",
    sprintf("%.3f", summary(aca.model.controls)$r.squared),
    sprintf("%d", nobs(aca.model.controls))
  ),
  check.names = FALSE
)

# Save to LaTeX
tex_file <- file.path(output_dir, "Exhibit_12_1_1A_r.tex")
print(xtable(results_table,
             caption = "ATEs With and Without Available Case Analysis",
             label = "tab:exhibit_12_1_1A"),
      file = tex_file,
      include.rownames = FALSE)
cat(sprintf("\n✓ Saved to: %s\n\n", tex_file))

# =============================================================================
# END OF EXHIBIT 12.1.1A
# =============================================================================
